File: GoL/Model/GameInitializer.m
//
//  GameInitializer.m
//  GoL
//
//  Created by Georg Meyer on 28.04.14.
//
//

#import "Cell.h"
#import "Dimensions.h"
#import "GameInitializer.h"
#import "NeighborhoodInitializer.h"

@interface GameInitializer ()

@property (strong, nonatomic) NSArray *initialState;
@property (strong, nonatomic) NSMutableArray *field;
@property (strong, nonatomic) NeighborhoodInitializer *neighborhoodInitializer;

@property (strong, nonatomic) Dimensions *dimensions;

@end


@implementation GameInitializer

- (id)initWithInitialState:(NSArray *)initialState andDimensions:(Dimensions *)dimensions {
    self = [super init];
    if (self) {
        self.initialState = initialState;
        self.dimensions = dimensions;
        self.field = [[NSMutableArray alloc] initWithCapacity:self.dimensions.height];

        self.neighborhoodInitializer = [[NeighborhoodInitializer alloc] initWithField:self.field andDimensions:self.dimensions];
    }
    return self;
}

- (NSArray *)initialize {
    [self setUpGame];
    return self.field;
}

- (void)setUpGame {
    for (NSArray *rowState in self.initialState) {
        [self setUpRow:rowState];
    }
}

- (void)setUpRow:(NSArray *)rowState {
    NSMutableArray *row = [[NSMutableArray alloc] initWithCapacity:self.dimensions.width];
    [self.field addObject:row];

    for (int i = 0; i < rowState.count; i++) {
        BOOL isAlive = [[rowState objectAtIndex:i] boolValue];
        [self setUpCell:isAlive];
    }
}

- (void)setUpCell:(BOOL)isAlive {
    Cell *cell = isAlive ? [Cell createAlive] : [Cell createDead];
    [self addCell:cell];

    [self.neighborhoodInitializer initializeCell:cell];
}

- (void)addCell:(Cell *)cell {
    [[self.field lastObject] addObject:cell];
}

@end

File: GoL/Model/GameInitializer.swift
//
//  GameInitializer.swift
//  GoL
//
//  Created by Georg Meyer on 08.06.14.
//
//

import Foundation


class GameInitializer {
    
    var initialState : Array<Array<Bool>>
    var field : Array<Array<Cell>>
    var neighborhoodInitializer : NeighborhoodInitializer
    var dimensions : Dimensions
    
    init(initialState : Array<Array<Bool>>, dimensions : Dimensions){
        self.initialState = initialState
        self.dimensions = dimensions
        self.field = Array<Array<Cell>>()
        self.neighborhoodInitializer = NeighborhoodInitializer(dimensions: dimensions)
        
    }
    
    func initialize() -> Array<Array<Cell>> {
        setUpGame()
        
        return field
    }
    
    func setUpGame() {
        for rowState in initialState {
            setUpRow(rowState)
        }
    }
    
    func setUpRow(rowState : Array<Bool>){
        var row : Array<Cell> = Array<Cell>()
        
        field.append(row)

        for isAlive in rowState {
            setUpCell(isAlive)
        }
    }
    
    func setUpCell(isAlive : Bool){
        let cell = isAlive ? Cell.createAlive() : Cell.createDead()
        addCell(cell)
        neighborhoodInitializer.initializeCell(cell, field: field)
    }
    
    func addCell(cell : Cell){
        self.field[self.field.count - 1].append(cell)
    }
}